datajoint_utilities/typing/__init__.py

```diff
@@ -0,0 +1,180 @@
+
+
+def is_strmap(obj: object, allow_empty: bool = True) -> TypeGuard[AnyMap]:
+    if not isinstance(obj, Mapping):
+        return False
+    if len(cast(Sized, obj)) == 0:
+        return allow_empty
+    return all(
+        isinstance(k, str) and isinstance(v, object)
+        for k, v in cast(ItemsView[Any, Any], obj.items())
+    )
+
+
+def is_uuid_str(obj: object) -> TypeGuard[UUIDStr]:
+    if isinstance(obj, UUID):
+        return True
+    elif isinstance(obj, str):
+        obj = obj.replace("urn:", "").replace("uuid:", "")
+        obj = obj.strip("{}").replace("-", "")
+        return len(obj) == 32
+    else:
+        return False
+
+
+def is_timestampable(obj: object) -> TypeGuard[TimeStampable]:
+    return isinstance(obj, (datetime.datetime, datetime.date, datetime.timedelta))
+
+
+@runtime_checkable
+class SupportsBool(Protocol):
+    def __bool__(self) -> bool:
+        ...
+
+
+# DataJoint typing ---------------------------------------------------------------------
+T_Table = TypeVar("T_Table", bound=dj.expression.QueryExpression)
+T_UserTable = TypeVar("T_UserTable", bound=dj.user_tables.UserTable)
+BaseTable: TypeAlias = dj.expression.QueryExpression
+UserTable: TypeAlias = dj.user_tables.UserTable
+AnyTable: TypeAlias = (
+    dj.expression.QueryExpression
+    | dj.table.Table
+    | dj.user_tables.UserTable
+    | dj.user_tables.TableMeta
+    | dj.user_tables.Lookup
+    | dj.user_tables.Manual
+    | dj.user_tables.Imported
+    | dj.user_tables.Computed
+    | dj.user_tables.Part
+)
+HeadingProp: TypeAlias = Literal["names", "primary_key", "secondary_attributes"]
+DBConnection: TypeAlias = dj.connection.Connection
+ContextLike: TypeAlias = AnyMap | FrameStack | types.ModuleType | str
+
+
+class InsertOpts(TypedDict, total=False):
+    replace: bool
+    skip_duplicates: bool
+    ignore_extra_fields: bool
+    allow_direct_insert: bool | None
+
+
+class InsertCallArgs(TypedDict):
+    row: DictObj
+    insert_opts: InsertOpts
+
+
+def is_djtable(obj: object) -> TypeGuard[dj.user_tables.UserTable]:
+    return isinstance(obj, BaseTable) or (
+        inspect.isclass(obj) and issubclass(obj, BaseTable)
+    )
+
+
+def is_parttable(obj: object) -> TypeGuard[dj.user_tables.Part]:
+    return inspect.isclass(obj) and issubclass(obj, dj.user_tables.Part)
+
+
+# Numpy typing -------------------------------------------------------------------------
+ArrayTypes: TypeAlias = bool_ | complex_ | float_ | int_ | str_
+VecTypes: TypeAlias = bool | complex | float | int | str
+NPArray: TypeAlias = NDArray[ArrayTypes]
+StrArray: TypeAlias = NDArray[str_]
+AnyArray: TypeAlias = NDArray[Any]
+Vector: TypeAlias = Sequence[VecTypes]
+ArrayOrVec: TypeAlias = (
+    NPArray
+    | Vector
+    | Sequence[Vector]
+    | Sequence[Sequence[Vector]]
+    | Sequence[Sequence[Sequence[Vector]]]
+    | Sequence[Sequence[Sequence[Sequence[Vector]]]]
+)
+
+
+def is_ndarray(obj: object) -> TypeGuard[AnyArray]:
```

Comment:
why is this function necessary?

Reply:
Purely for the mypy checker to be able to narrow types in conditionals